test/unit/security-rules/security-rules.spec.ts

```diff
@@ -377,6 +396,213 @@ describe('SecurityRules', () => {
     });
   });
 
+  describe('releaseFirestoreRulesetFromSource', () => {
+    const RULES_FILE = {
+      name: 'firestore.rules',
+      content: 'test source {}',
+    };
+
+    INVALID_SOURCES.forEach((invalidSource) => {
+      it(`should reject when called with: ${JSON.stringify(invalidSource)}`, () => {
+        return securityRules.releaseFirestoreRulesetFromSource(invalidSource)
+          .should.eventually.be.rejected.and.deep.equal(INVALID_SOURCE_ERROR);
+      });
+    });
+
+    it('should propagate API errors', () => {
+      const stub = sinon
+        .stub(SecurityRulesApiClient.prototype, 'createRuleset')
+        .rejects(EXPECTED_ERROR);
+      stubs.push(stub);
+      return securityRules.releaseFirestoreRulesetFromSource('foo')
+        .should.eventually.be.rejected.and.deep.equal(EXPECTED_ERROR);
+    });
+
+    const sources: {[key: string]: string | Buffer} = {
+      string: RULES_FILE.content,
+      buffer: Buffer.from(RULES_FILE.content),
+    };
+    Object.keys(sources).forEach((key) => {
+      it(`should resolve on success when source specified as a ${key}`, () => {
+        const [createRuleset, updateRelease] = stubReleaseFromSource();
+
+        return securityRules.releaseFirestoreRulesetFromSource(sources[key])
+          .then((ruleset) => {
+            expect(ruleset.name).to.equal('foo');
+            expect(ruleset.createTime).to.equal(CREATE_TIME_UTC);
+            expect(ruleset.source.length).to.equal(1);
+
+            const file = ruleset.source[0];
+            expect(file.name).equals('firestore.rules');
+            expect(file.content).equals('service cloud.firestore{\n}\n');
```

Comment:
I do not think this is a valid ruleset, where is it coming from? I see `RULES_FILE` above has content `test source {}`, does this test pass?

Reply:
These are unit tests. They pass because they don't make any actual RPC calls.